File: starship_duel/arena/sdk/python/example_bot.py
#!/usr/bin/env python3
"""Example Starship Duel bot (Python). Run it against the engine with:

    python -m starship_duel.run --bot0 heuristic \
        --bot1 "cmd:python starship_duel/arena/sdk/python/example_bot.py"

It keeps a tiny belief of its own across turns and plays a simple hunt-and-hide
strategy — a starting point to build on.
"""

from starship_sdk import run

# Per-game memory (the process persists for the whole game).
_last_known = None


def decide(req: dict) -> dict:
    global _last_known
    me = req["you"]
    rival = req["rival"]
    legal = req["legal_actions"]
    types = {a["action"] for a in legal}

    known = rival["known_position"]
    if known:
        _last_known = known

    # 1) Kill if the rival is confirmed in our system.
    if known == me["position"] and "FIRE" in types:
        return {"action": "FIRE"}

    # 2) If the rival is known adjacent and we can jump+fire this turn, pounce.
    if known and me["actions_remaining"] >= 2:
        for a in legal:
            if a["action"] == "JUMP" and a.get("target") == known:
                return a

    # 3) If exposed, slip back under cloak.
    if not me["cloaked"] and "HOLD" in types:
        return {"action": "HOLD"}

    # 4) Grab an unclaimed binary we're standing on for income.
    here = req["systems"].get(me["position"], {})
    if "CLAIM" in types and here.get("binary") and here.get("owner") is None:
        return {"action": "CLAIM"}

    # 5) Otherwise move toward a binary, avoiding collapsing systems.
    jumps = [a for a in legal if a["action"] == "JUMP"
             and req["systems"].get(a["target"], {}).get("status") == "STABLE"]
    if jumps:
        # prefer a binary destination if one is adjacent
        for a in jumps:
            if req["systems"].get(a["target"], {}).get("binary"):
                return a
        return jumps[0]

    return {"action": "HOLD" if "HOLD" in types else "END_TURN"}


if __name__ == "__main__":
    run(decide)

File: starship_duel/arena/sdk/cpp/example_bot.cpp
// Example Starship Duel bot (C++). Build & run:
//
//   g++ -std=c++17 -O2 example_bot.cpp -o example_bot        # needs nlohmann/json
//   python -m starship_duel.run --bot0 heuristic --bot1 "cmd:./example_bot"
//
// Simple hunt-and-hide strategy, mirroring the Python example.

#include "starship_bot.hpp"
using nlohmann::json;

// Per-game memory persists for the whole process.
static std::string g_last_known;

static bool has_action(const json& legal, const std::string& name) {
    for (const auto& a : legal)
        if (a["action"] == name) return true;
    return false;
}

json decide(const json& req) {
    const auto& me = req["you"];
    const auto& rival = req["rival"];
    const auto& legal = req["legal_actions"];
    const auto& systems = req["systems"];
    std::string pos = me["position"].get<std::string>();

    std::string known = rival["known_position"].is_string()
                        ? rival["known_position"].get<std::string>() : "";
    if (!known.empty()) g_last_known = known;

    // 1) Kill if the rival is confirmed in our system.
    if (known == pos && has_action(legal, "FIRE"))
        return json{{"action", "FIRE"}};

    // 2) Rival known adjacent + enough actions -> jump on (fire next).
    if (!known.empty() && me["actions_remaining"].get<int>() >= 2)
        for (const auto& a : legal)
            if (a["action"] == "JUMP" && a.value("target", "") == known)
                return a;

    // 3) Exposed -> re-cloak.
    if (!me["cloaked"].get<bool>() && has_action(legal, "HOLD"))
        return json{{"action", "HOLD"}};

    // 4) Standing on an unclaimed binary -> claim it.
    if (has_action(legal, "CLAIM") && systems.contains(pos)
        && systems[pos].value("binary", false) && systems[pos]["owner"].is_null())
        return json{{"action", "CLAIM"}};

    // 5) Move toward a stable binary, else any stable system.
    const json* fallback = nullptr;
    for (const auto& a : legal) {
        if (a["action"] != "JUMP") continue;
        std::string tgt = a.value("target", "");
        if (!systems.contains(tgt) || systems[tgt].value("status", "") != "STABLE") continue;
        if (systems[tgt].value("binary", false)) return a;
        if (!fallback) fallback = &a;
    }
    if (fallback) return *fallback;

    return json{{"action", has_action(legal, "HOLD") ? "HOLD" : "END_TURN"}};
}

int main() { return starship::run(decide); }
